akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -2285,6 +2285,281 @@ private[akka] final class StatefulMap[S, In, Out](
   override def toString = "StatefulMap"
 }
 
+/**
+ * INTERNAL API
+ */
+private[akka] final case class StatefulMapAsync[S, In, Out](parallelism: Int)(
+    create: () => Future[S],
+    //TODO (S, In) => (S, Future[Out]) seems more comfortable for user
+    f: (S, In) => Future[(S, Out)],
+    onComplete: S => Future[Option[Out]],
+    combineState: (S, S) => S)
+    extends GraphStage[FlowShape[In, Out]] {
+
+  import MapAsync._
+
+  private val in = Inlet[In]("StatefulMapAsync.in")
+  private val out = Outlet[Out]("StatefulMapAsync.out")
+
+  override def initialAttributes = DefaultAttributes.statefulMapAsync and SourceLocation.forLambda(f)
+
+  override val shape = FlowShape(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with InHandler with OutHandler {
+
+      setHandlers(in, out, this)
+
+      lazy val decider: Decider = inheritedAttributes.mandatoryAttribute[SupervisionStrategy].decider
+
+      // if release hasn't been called
+      private var stateAcquired: Boolean = true
+      // if it is need to push element after state initialized
+      // we need this flag bcs if stream ever been pushed or pulled at the time state doesn't available
+      // that will require triggering push in create callback, but prevent push been triggered by just preStart
+      private var tryPushAfterInitialized: Boolean = false
+      private var state: Option[S] = None
```

Comment:
OptionVal or just `S`?

Reply:
trying to avoid using plain null here.
i guess little overhead doesn't matter here? but nice to have this optimization anyway.
i will replace it later.